src/julia.h

```diff
@@ -1462,16 +1493,52 @@ JL_DLLEXPORT JL_CONST_FUNC jl_tls_states_t *(jl_get_ptls_states)(void);
 #ifndef JULIA_ENABLE_THREADING
 extern JL_DLLEXPORT jl_tls_states_t jl_tls_states;
 #define jl_get_ptls_states() (&jl_tls_states)
-#else
+STATIC_INLINE int8_t jl_gc_state_set(int8_t state, int8_t old_state)
+{
+    (void)state;
+    return old_state;
+}
+STATIC_INLINE int8_t jl_gc_state_save_and_set(int8_t state)
+{
+    (void)state;
+    return 0;
+}
+#define jl_gc_unsafe_enter() jl_gc_state_save_and_set(0)
+#define jl_gc_unsafe_leave(state) ((void)state)
+#define JL_RETURN_NULL_IF_GCSAFE() do {} while (0)
+#else // ifndef JULIA_ENABLE_THREADING
 typedef jl_tls_states_t *(*jl_get_ptls_states_func)(void);
 JL_DLLEXPORT void jl_set_ptls_states_getter(jl_get_ptls_states_func f);
-#endif
+STATIC_INLINE int8_t jl_gc_state_set(int8_t state, int8_t old_state)
+{
+    jl_get_ptls_states()->gc_state = state;
+    // A safe point is required if we transition from GC-safe region to
+    // non GC-safe region.
+    if (old_state && !state)
+        jl_gc_safepoint();
+    return old_state;
+}
+STATIC_INLINE int8_t jl_gc_state_save_and_set(int8_t state)
+{
+    return jl_gc_state_set(state, jl_get_ptls_states()->gc_state);
+}
+#define jl_gc_unsafe_enter() jl_gc_state_save_and_set(0)
+#define jl_gc_unsafe_leave(state) jl_gc_state_set((state), 0)
+// Return NULL when we are in GC safe mode and the return value can be free'd
+// by GC on another thread
+#define JL_RETURN_NULL_IF_GCSAFE() do {          \
```

Comment:
Also done. I also don't really like `return` inside a macro although I sometimes treat error checking macros with a `return` in the name as the only exception.

I used a macro here since I want to make sure the non-threading version has neither the volatile load and the branch. I just realized that I could just define `jl_gc_state` as `0` on non-threading and branch on that for `NULL` return.